File: ScriptGraph/TemplateScriptNode/PalmUp/CGPalmUp.lua
local CGPalmUp = CGPalmUp or {}
CGPalmUp.__index = CGPalmUp

function CGPalmUp.new()
    local self = setmetatable({}, CGPalmUp)
    self.inputs = {}
    self.nexts = {}
    self.handAction = ""
    return self
end

function CGPalmUp:setNext(index, func)
    self.nexts[index] = func
end

function CGPalmUp:setInput(index, func)
    self.inputs[index] = func
end

function CGPalmUp:update()
    if self.nexts[0] == nil and self.nexts[1] == nil then
        return
    end

    local result = Amaz.Algorithm.getAEAlgorithmResult()
    local hand = result:getHandInfo(self.inputs[0]())
    if hand == nil then
        return
    end

    local has_action = hand.action == Amaz.HandAction.PLAM_UP
    if self.handAction ~= "PALM_UP" and has_action then
        self.handAction = "PALM_UP"
        if self.nexts[0] then
            self.nexts[0]()
        end
    elseif self.handAction == "PALM_UP" and not has_action then
        self.handAction = ""
        if self.nexts[1] then
            self.nexts[1]()
        end
    end
end
return CGPalmUp

File: ScriptGraph/TemplateScriptNode/PalmUp/CGPalmUp.js
/**
 * @file CGPalmUp.js
 * @author liujiacheng
 * @date 2021/8/18
 * @brief CGPalmUp.js
 * @copyright Copyright (c) 2021, ByteDance Inc, All Rights Reserved
 */

const {BaseNode} = require('./BaseNode');
const Amaz = effect.Amaz;

class CGPalmUp extends BaseNode {
  constructor() {
    super();
    this.handAction = '';
  }

  setNext(index, func) {
    this.nexts[index] = func;
  }

  setInput(index, func) {
    this.inputs[index] = func;
  }

  onUpdate(sys, dt) {
    let v1 = this.inputs[0]();

    const algMgr = Amaz.AmazingManager.getSingleton('Algorithm');
    const algResult = algMgr.getAEAlgorithmResult();

    if (algResult && v1 != null) {
      const hand = algResult.getHandInfo(v1);
      if (hand == undefined || hand == null) {
        return;
      }

      const has_action = hand.action == Amaz.HandAction.PLAM_UP;
      if (has_action && this.handAction != 'PULM_UP') {
        this.handAction = 'PULM_UP';
        console.log('detect palm up!');
        if (this.nexts[0]) {
          this.nexts[0]();
        }
      } else if (!has_action && this.handAction == 'PULM_UP') {
        this.handAction = '';
        if (this.nexts[1]) {
          this.nexts[1]();
        }
      }
    }
  }
}

exports.CGPalmUp = CGPalmUp;
